cpp/src/arrow/flight/sql/client.h

```diff
@@ -135,6 +135,19 @@ class ARROW_EXPORT FlightSqlClient {
   arrow::Result<std::unique_ptr<FlightInfo>> GetTableTypes(
       const FlightCallOptions& options);
 
+  /// \brief Request the information about all the data types supported.
+  /// \param[in] options          RPC-layer hints for this call.
+  /// \return The FlightInfo describing where to access the dataset.
+  arrow::Result<std::unique_ptr<FlightInfo>> GetTypeInfo(
+      const FlightCallOptions& options);
+
+  /// \brief Request the information about all the data types supported.
+  /// \param[in] options          RPC-layer hints for this call.
+  /// \param[in] data_type        The data type to search for as filtering.
+  /// \return The FlightInfo describing where to access the dataset.
+  arrow::Result<std::unique_ptr<FlightInfo>> GetTypeInfo(const FlightCallOptions& options,
+                                                         int data_type);
```

Comment:
Are there situations where it is desirable to call this instead of getting all data types supported (there shouldn't be thousands of them...)?

Reply:
For the JDBC use no, since it always return all the typeInfo. However, when the ODBC start using this implementaion, it will require a filter. 

The ODBC call for the sqlType info takes a parameter

**e.g.**
When calling like this, it will return all the types. In this case the ODBC layer would call the one without filtering.
```
      result = SQLGetTypeInfo(g_stmt, SQL_ALL_TYPES);
```

If you want to filter you should call like this. Then, the ODBC layer would call the one that takes the filter parameter
```
      result = SQLGetTypeInfo(g_stmt, SQL_INTEGER);
```